File: yousto/ios/YoustoCalendar.m
// YoustoCalendar.m
// Programmatically adds events to iOS Calendar using EKEventStore.
// Works on both Simulator and real device — no EKEventEditViewController needed.
// Requests write-only access (iOS 17+) or full access (iOS ≤16).
// Returns eventIdentifier on success for future Calendar deep-linking.
//
// Pure ObjC — no Swift, no bridging header required.

#import <React/RCTBridgeModule.h>
@import EventKit;
@import UIKit;

// ── Module ────────────────────────────────────────────────────────────────────

@interface YoustoCalendar : NSObject <RCTBridgeModule>
@end

@implementation YoustoCalendar

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(presentEventCreationDialog:(NSDictionary *)eventData
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {

  dispatch_async(dispatch_get_main_queue(), ^{
    EKEventStore *store = [[EKEventStore alloc] init];

    // ── Permission request ────────────────────────────────────────────────────
    // Use write-only access on iOS 17+ (least-privilege principle),
    // fall back to full access on older OS versions.

    void (^addEventBlock)(void) = ^{
      // ── Build the EKEvent ───────────────────────────────────────────────────
      EKEvent *event = [EKEvent eventWithEventStore:store];

      // Title
      event.title = eventData[@"title"] ?: @"Event";

      // Start date — parse ISO-8601 string from JS
      NSString *dateStr = eventData[@"startDate"];
      if (dateStr && [dateStr isKindOfClass:[NSString class]] && dateStr.length > 0) {
        // Try ISO8601 formatter first (handles Z and +HH:MM offsets)
        NSISO8601DateFormatter *isoFmt = [[NSISO8601DateFormatter alloc] init];
        NSDate *parsed = [isoFmt dateFromString:dateStr];

        // Fallback: plain "yyyy-MM-dd'T'HH:mm:ss" without timezone
        if (!parsed) {
          NSDateFormatter *plain = [[NSDateFormatter alloc] init];
          plain.locale   = [NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"];
          plain.timeZone = [NSTimeZone localTimeZone];
          plain.dateFormat = @"yyyy-MM-dd'T'HH:mm:ss";
          parsed = [plain dateFromString:dateStr];
        }

        if (parsed) {
          event.startDate = parsed;
          event.endDate   = [parsed dateByAddingTimeInterval:3600]; // 1-hour block
          event.allDay    = NO;
        }
      }

      // If no valid date was parsed, default to tomorrow all-day
      if (!event.startDate) {
        NSCalendar *cal = [NSCalendar currentCalendar];
        NSDateComponents *comps = [NSDateComponents new];
        comps.day = 1;
        NSDate *tomorrow = [cal dateByAddingComponents:comps toDate:[NSDate date] options:0];
        event.startDate = tomorrow;
        event.endDate   = tomorrow;
        event.allDay    = YES;
      }

      // Notes / description
      NSString *notes = eventData[@"notes"];
      if (notes && [notes isKindOfClass:[NSString class]] && notes.length > 0) {
        event.notes = notes;
      }

      // Location
      NSString *location = eventData[@"location"];
      if (location && [location isKindOfClass:[NSString class]] && location.length > 0) {
        event.location = location;
      }

      // Calendar — write to the user's default calendar
      event.calendar = store.defaultCalendarForNewEvents;

      // ── Save ────────────────────────────────────────────────────────────────
      NSError *saveError = nil;
      BOOL saved = [store saveEvent:event span:EKSpanThisEvent error:&saveError];

      if (!saved || saveError) {
        reject(@"SAVE_FAILED",
               saveError.localizedDescription ?: @"Failed to save event to Calendar",
               saveError);
        return;
      }

      NSString *eventId = event.eventIdentifier ?: @"";
      NSLog(@"[YoustoCalendar] Event saved: %@ (id: %@)", event.title, eventId);
      resolve(@{ @"action": @"saved", @"eventIdentifier": eventId });
    };

    // ── iOS 17+ write-only access ─────────────────────────────────────────────
    if (@available(iOS 17.0, *)) {
      [store requestWriteOnlyAccessToEventsWithCompletion:^(BOOL granted, NSError *error) {
        if (!granted || error) {
          reject(@"PERMISSION_DENIED",
                 error.localizedDescription ?: @"Calendar write permission denied",
                 error);
          return;
        }
        dispatch_async(dispatch_get_main_queue(), addEventBlock);
      }];
    } else {
      // ── iOS ≤16 full access ─────────────────────────────────────────────────
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
      [store requestAccessToEntityType:EKEntityTypeEvent completion:^(BOOL granted, NSError *error) {
        if (!granted || error) {
          reject(@"PERMISSION_DENIED",
                 error.localizedDescription ?: @"Calendar permission denied",
                 error);
          return;
        }
        dispatch_async(dispatch_get_main_queue(), addEventBlock);
      }];
#pragma clang diagnostic pop
    }
  });
}

@end

File: yousto/ios/YoustoCalendar.swift
// YoustoCalendar.swift
// Native module that presents EKEventEditViewController — the native iOS calendar
// event creation dialog — with pre-filled event data from Yousto.
//
// No calendar permissions are requested in advance; iOS handles permission
// implicitly when the user taps "Add" in the native dialog.
//
// ─── Xcode setup ───────────────────────────────────────────────────────────
// 1. In Xcode, File > Add Files to "YoustoApp" > select this file
// 2. Also add YoustoCalendar.m to the same target
// 3. If a YoustoApp-Bridging-Header.h does not exist, Xcode will offer to
//    create one automatically when you add the .m file — accept it.
// 4. Rebuild (Cmd+R). NativeModules.YoustoCalendar will then be defined.

import Foundation
import EventKit
import EventKitUI
import UIKit

// Retain the delegate for the lifetime of the presented view controller.
private var calendarDelegateKey = "calendarDelegate"

@objc(YoustoCalendar)
class YoustoCalendar: NSObject {

  // React Native calls module methods on a background thread by default.
  // We need the main queue for UIKit operations.
  @objc static func requiresMainQueueSetup() -> Bool { return false }

  @objc
  func presentEventCreationDialog(
    _ eventData: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      let store = EKEventStore()
      let event = EKEvent(eventStore: store)

      // ── Title ────────────────────────────────────────────────────────────
      event.title = eventData["title"] as? String ?? "Event"

      // ── Start date ───────────────────────────────────────────────────────
      if let dateStr = eventData["startDate"] as? String {
        // Try ISO 8601 with timezone offset first (most common from JS Date.toISOString())
        let isoFormatter = ISO8601DateFormatter()
        isoFormatter.formatOptions = [
          .withInternetDateTime,
          .withDashSeparatorInDate,
          .withColonSeparatorInTime,
          .withTimeZone,
        ]
        if let date = isoFormatter.date(from: dateStr) {
          event.startDate = date
        } else {
          // Fallback: plain datetime without timezone (e.g. "2025-03-18T09:00:00")
          let fallbackFormatter = DateFormatter()
          fallbackFormatter.locale = Locale(identifier: "en_US_POSIX")
          fallbackFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
          event.startDate = fallbackFormatter.date(from: dateStr) ?? Date().addingTimeInterval(86400)
        }
      } else {
        // No date extracted — default to tomorrow at 9 AM
        var components = Calendar.current.dateComponents([.year, .month, .day], from: Date())
        components.day = (components.day ?? 1) + 1
        components.hour = 9
        components.minute = 0
        event.startDate = Calendar.current.date(from: components) ?? Date().addingTimeInterval(86400)
      }

      // Default 1-hour duration
      event.endDate = event.startDate.addingTimeInterval(3600)

      // ── Notes / description ───────────────────────────────────────────────
      if let notes = eventData["notes"] as? String, !notes.isEmpty {
        event.notes = notes
      }

      // ── Present EKEventEditViewController ─────────────────────────────────
      let editVC = EKEventEditViewController()
      editVC.event = event
      editVC.eventStore = store

      // Create delegate and retain it on the view controller via associated object
      let delegate = CalendarEditDelegate(resolve: resolve, reject: reject)
      editVC.editViewDelegate = delegate
      objc_setAssociatedObject(
        editVC,
        &calendarDelegateKey,
        delegate,
        .OBJC_ASSOCIATION_RETAIN_NONATOMIC
      )

      // Find the top-most presented view controller to present from
      guard let scene = UIApplication.shared.connectedScenes
              .compactMap({ $0 as? UIWindowScene })
              .first(where: { $0.activationState == .foregroundActive }),
            let rootVC = scene.windows.first(where: { $0.isKeyWindow })?.rootViewController
      else {
        reject("NO_ROOT_VC", "Could not find root view controller", nil)
        return
      }

      var topVC = rootVC
      while let presented = topVC.presentedViewController {
        topVC = presented
      }

      topVC.present(editVC, animated: true, completion: nil)
    }
  }
}

// ── EKEventEditViewDelegate ────────────────────────────────────────────────────
// Captures the result when the user taps "Add", "Cancel", or "Delete".

private class CalendarEditDelegate: NSObject, EKEventEditViewDelegate {
  private let resolve: RCTPromiseResolveBlock
  private let reject: RCTPromiseRejectBlock

  init(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    self.resolve = resolve
    self.reject = reject
  }

  func eventEditViewController(
    _ controller: EKEventEditViewController,
    didCompleteWith action: EKEventEditViewAction
  ) {
    controller.dismiss(animated: true, completion: nil)

    switch action {
    case .saved:
      resolve(["action": "saved"])
    case .canceled:
      resolve(["action": "canceled"])
    case .deleted:
      resolve(["action": "deleted"])
    @unknown default:
      resolve(["action": "unknown"])
    }
  }
}
